debian/control

```diff
@@ -0,0 +1,37 @@
+Source: ultrastardx
+Section: games
+Priority: optional
+Maintainer: Markus Meier <memcopy@epic-bytes.net>
+Build-Depends: debhelper (>= 10),
+ autotools-dev,
+ fpc (>= 3.0.0),
+ libpng-dev,
+ libfreetype6-dev,
+ libsqlite3-dev,
+ libsdl2-dev,
+ libsdl2-image-dev,
+ liblua5.3-dev,
+ libavcodec-dev,
+ libavformat-dev,
+ libavutil-dev,
+ portaudio19-dev,
+ libswscale-dev,
+ libprojectm-dev,
+ libpcre3-dev
+Standards-Version: 3.9.8
+Homepage: https://usdx.eu
+
+Package: ultrastardx
+Architecture: any
+Depends: ${shlibs:Depends}, ${misc:Depends},
```

Comment:
libGL.so and libGLU.so are dlopen'ed. They are in libgl1-mesa-dev and libglu1-mesa-dev.

Reply:
I have non of these installed and USDX seems to run flawlessly. I'm confused now. Does it mean I run USDX without OpenGL on my system?

Reply:
USDX will not work without OpenGL. You must have a libGL.so somewhere.
Try `strace -e open -o logfile ./ultrastardx ; grep libGL logfile`.

At least for libGL it should be possible to rewrite dglGetProcAddress to use SDL_GL_GetProcAddress and skip the dglLoadLibrary step.

Reply:
Nope, here is the proof. Look for the last lines:
open("/usr/lib/x86_64-linux-gnu/libGLEW.so.2.0", O_RDONLY|O_CLOEXEC) = 3
open("/usr/lib/x86_64-linux-gnu/libGLU.so.1", O_RDONLY|O_CLOEXEC) = 3
open("/usr/lib/x86_64-linux-gnu/libGL.so.1", O_RDONLY|O_CLOEXEC) = 3
open("/usr/lib/x86_64-linux-gnu/libGLX.so.0", O_RDONLY|O_CLOEXEC) = 3
open("/usr/lib/x86_64-linux-gnu/libGLdispatch.so.0", O_RDONLY|O_CLOEXEC) = 3
open("/usr/lib/x86_64-linux-gnu/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7
open("/usr/lib/x86_64-linux-gnu/libGL.so", O_RDONLY|O_CLOEXEC) = 32
open("/lib/x86_64-linux-gnu/libGLU.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/x86_64-linux-gnu/libGLU.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/lib/libGLU.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/libGLU.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)

It looks like SDL is initializing the OpenGL context and then everything is working despite the missing libGL.so and libGLU.so symlinks.

Reply:
Ok, it works for the following reasons:
- ReadOpenGLCore and ReadExtensions don't check if GL_LibHandle is set before they call dglGetProcAddress
- RTLD_DEFAULT == nil
- Some other library already loaded libGL.so.1
- We don't need the functions pointers that are initialized only if GL_LibHandle <> nil (i.e. glX*)
- We don't use dglOpenGL to access GLU. We use Free Pascal's glu unit, which dlopen's libGLU.so.1.